src/topology/metric_space/emetric_space.lean

```diff
@@ -729,40 +729,64 @@ end second_countable
 section diam
 
 /-- The diameter of a set in an emetric space, named `emetric.diam` -/
-def diam (s : set α) := Sup ((λp : α × α, edist p.1 p.2) '' (set.prod s s))
+def diam (s : set α) := ⨆ (x ∈ s) (y ∈ s), edist x y
+
+lemma diam_le_iff_forall_edist_le {d : ennreal} :
+  diam s ≤ d ↔ ∀ (x ∈ s) (y ∈ s), edist x y ≤ d :=
+by simp only [diam, supr_le_iff]
 
 /-- If two points belong to some set, their edistance is bounded by the diameter of the set -/
 lemma edist_le_diam_of_mem (hx : x ∈ s) (hy : y ∈ s) : edist x y ≤ diam s :=
-le_Sup ((mem_image _ _ _).2 ⟨(⟨x, y⟩ : α × α), by simp [hx, hy]⟩)
+diam_le_iff_forall_edist_le.1 (le_refl _) x hx y hy
 
 /-- If the distance between any two points in a set is bounded by some constant, this constant
 bounds the diameter. -/
-lemma diam_le_of_forall_edist_le {d : ennreal} (h : ∀x y ∈ s, edist x y ≤ d) : diam s ≤ d :=
-begin
-  apply Sup_le _,
-  simp only [and_imp, set.mem_image, set.mem_prod, exists_imp_distrib, prod.exists],
-  assume b x y xs ys dxy,
-  rw ← dxy,
-  exact h x y xs ys
-end
+lemma diam_le_of_forall_edist_le {d : ennreal} (h : ∀ (x ∈ s) (y ∈ s), edist x y ≤ d) :
+  diam s ≤ d :=
+diam_le_iff_forall_edist_le.2 h
+
+/-- The diameter of a subsingleton vanishes. -/
+lemma diam_subsingleton (hs : s.subsingleton) : diam s = 0 :=
+le_zero_iff_eq.1 $ diam_le_of_forall_edist_le $
+λ x hx y hy, (hs hx hy).symm ▸ edist_self y ▸ le_refl _
 
 /-- The diameter of the empty set vanishes -/
 @[simp] lemma diam_empty : diam (∅ : set α) = 0 :=
-by simp [diam]
+diam_subsingleton subsingleton_empty
 
 /-- The diameter of a singleton vanishes -/
 @[simp] lemma diam_singleton : diam ({x} : set α) = 0 :=
-by simp [diam]
+diam_subsingleton subsingleton_singleton
 
-/-- The diameter is monotonous with respect to inclusion -/
-lemma diam_mono {s t : set α} (h : s ⊆ t) : diam s ≤ diam t :=
+lemma diam_eq_zero_iff : diam s = 0 ↔ s.subsingleton :=
+⟨λ h x y hx hy, edist_le_zero.1 $ h ▸ edist_le_diam_of_mem hx hy, diam_subsingleton⟩
+
+lemma diam_pos_iff : 0 < diam s ↔ ∃ x y ∈ s, x ≠ y :=
```

Comment:
For coherence with the other modifications you made, I would also use `(x ∈ s) (y ∈ s)` instead of `x y ∈ s` here. (And I agree your choice is better, as there is less potential confusion).

Reply:
Done; redefined `set.subsingleton` in a similar way.